File: goldenflow/engine/differ.py
from __future__ import annotations

from dataclasses import dataclass, field

import polars as pl


@dataclass
class DiffResult:
    total_changes: int = 0
    changed_columns: list[str] = field(default_factory=list)
    added_columns: list[str] = field(default_factory=list)
    removed_columns: list[str] = field(default_factory=list)
    row_count_before: int = 0
    row_count_after: int = 0
    column_details: dict[str, dict] = field(default_factory=dict)


def diff_dataframes(before: pl.DataFrame, after: pl.DataFrame) -> DiffResult:
    """Compare two DataFrames and report differences."""
    result = DiffResult(
        row_count_before=before.shape[0],
        row_count_after=after.shape[0],
    )

    before_cols = set(before.columns)
    after_cols = set(after.columns)

    result.added_columns = sorted(after_cols - before_cols)
    result.removed_columns = sorted(before_cols - after_cols)

    common_cols = before_cols & after_cols
    total_changes = 0

    for col in sorted(common_cols):
        if before.shape[0] != after.shape[0]:
            result.changed_columns.append(col)
            total_changes += abs(before.shape[0] - after.shape[0])
            continue

        try:
            b_series = before[col].cast(pl.Utf8)
            a_series = after[col].cast(pl.Utf8)
            changes = (b_series != a_series).sum()
            if changes > 0:
                result.changed_columns.append(col)
                total_changes += changes
                result.column_details[col] = {"changed_rows": changes}
        except Exception:
            result.changed_columns.append(col)
            total_changes += before.shape[0]

    result.total_changes = total_changes
    return result

File: packages/goldenflow-js/src/core/engine/differ.ts
/**
 * Differ — compare two row arrays and report differences.
 */

import type { DiffResult, Row } from "../types.js";

export function diffDataframes(
  before: readonly Row[],
  after: readonly Row[],
): DiffResult {
  const beforeCols = new Set(before.length > 0 ? Object.keys(before[0]!) : []);
  const afterCols = new Set(after.length > 0 ? Object.keys(after[0]!) : []);

  const addedColumns = [...afterCols].filter((c) => !beforeCols.has(c)).sort();
  const removedColumns = [...beforeCols].filter((c) => !afterCols.has(c)).sort();
  const commonCols = [...beforeCols].filter((c) => afterCols.has(c)).sort();

  const changedColumns: string[] = [];
  const columnDetails: Record<string, { changedRows: number }> = {};
  let totalChanges = 0;

  for (const col of commonCols) {
    if (before.length !== after.length) {
      changedColumns.push(col);
      totalChanges += Math.abs(before.length - after.length);
      continue;
    }

    let changes = 0;
    for (let i = 0; i < before.length; i++) {
      const bVal = String(before[i]![col] ?? "");
      const aVal = String(after[i]![col] ?? "");
      if (bVal !== aVal) changes++;
    }

    if (changes > 0) {
      changedColumns.push(col);
      totalChanges += changes;
      columnDetails[col] = { changedRows: changes };
    }
  }

  return {
    totalChanges,
    changedColumns,
    addedColumns,
    removedColumns,
    rowCountBefore: before.length,
    rowCountAfter: after.length,
    columnDetails,
  };
}
